### ISSUE TITLE
Diagnostics show up on non-Mapper types

### ISSUE THREAD

**Issue by agudian** [mapstruct/mapstruct#598]
Suppose I have these types:

``` java
public interface BaseEntityMapper<E extends AbstractEntity, D extends AbstractDto> {
    E toEntity(D src);
    D toDto(E src);
}

@Mapper
public interface CarEntityMapper extends BaseEntityMapper<Car, CarDto> {
// nothing
}
```

If Car and CarDto map nicely, then `CarEntityMapperImpl` will be generated as one would expect and it works great. 

However, if they don't map nicely, e.g. due to unmapped target properties, the diagnostics are generated with the locations of the methods in BaseEntityMapper. If that type is located in an already created jar, the marker might not be visible in some IDEs or in javac.

I think that, generally, if we create a diagnostic for some mapper to generate, we should only use locations within that mapper. In this case, an "unmapped target property" could be located at "extends _BaseEntityMapper_".

WDYT?

**Comment by gunnarmorling** [comment]
Yes, I think that makes sense. I didn't foresee that usage pattern, but it seems legit and as you say all warnings should be raised at the `@Mapper`-annotated class. The warning would have to provide some more context though, e.g. the name of the affected method.

**Comment by Zegveld** [comment]
Thought I'd take my attempt at creating a fix for this issue. 

My idea (which I'm already implementing as you might have noticed by the commit above) is to pass both the MapperElement containing the sourceMethod and the sourceMethodElement itself to the FormattingMessager. There the FormattingMessager will check if the MapperElement contains the sourceMethodElement.
If it does then it will just report it as it currently already does.
If it does not then it will report it as coming from the mapper. Still need to add a bit of message text to display which method it actually originated from.
Currently thinking of adding something like `Error occured at '<Method>' located in '<MethodClass>'.` at the end of the original message.

Also it allows the use of null for the mapper element, because currently I've only adjusted the errors that appear at Method level. This change might impact other PR's because it changes the FormattingMessager interface.

----------------------

A larger example to explain the requested feature: (and a check to see if I understood it correctly)
```
@Mapper
interface Mapper extends BaseMapper<SourceClass, TargetClass> {
    public TargetClass myMap(SourceClass src);
}

interface BaseMapper<S, T> {
    public T map(S s);
}

class SourceClass {
    public String property;
}

class TargetClass {
    public Object property;
}
```
Should throw the exception at the `@Mapper` annotation/annotated class for the map method, but not for the `myMap` method. The `myMap` method should get the exception thrown at its own location.

----------------------

Open for suggestions/idea's/other-routes concerning the implementation of this feature.

**Comment by Zegveld** [comment]
Interesting note to add:
when trying to add the message I get different results depending on the compiler:
jdk: `Occured at 'map(SOURCE)' in 'AbstractMapper'.`
eclipse: `Occured at 'public abstract TARGET map(SOURCE) ' in 'AbstractMapper'.`

**Comment by Zegveld** [comment]
This feature request has been implemented in PR #2634. Closing this issue now.
